### PR TITLE
Redirect to home page when an invalid URL or missing document

### PR DESCRIPTION
**Pull request by liangshiw**
Resolved : https://github.com/abpframework/abp/issues/3637

### CODE REVIEW

**Comment by maliming** [comment]
Maybe we can show an error page before redirecting, and then redirect automatically after a few seconds. This allows the user to know what happened.

**Comment by liangshiw** [comment]
Ok.

File: modules/docs/src/Volo.Docs.Domain/Volo/Docs/Localization/Domain/zh-Hans.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
     "Projects": "项目",
     "NoProjectWarning": "还没有项目!",
     "DocumentNotFound": "找不到请求的文档!",
+    "ProjectNotFound": "找不到請求的項目!",
```
Comment (Rwing) [inline modules/docs/src/Volo.Docs.Domain/Volo/Docs/Localization/Domain/zh-Hans.json:17]: Is it Traditional Chinese?

**Comment by maliming** [comment]
@liangshiw Are you done?

**Comment by liangshiw** [comment]
@maliming  Yes ,all done.

File: modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/Index.cshtml.cs [hunk 2]
```diff
@@ -36,6 +34,10 @@ public class IndexModel : AbpPageModel
         public string LanguageCode { get; set; }
 
         public bool DocumentFound { get; set; } = true;
+        
+        public bool EntityFound { get; set; } = true;
```
Comment (maliming) [inline modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/Index.cshtml.cs:38]: It should be renamed to `ProjectFound`.

`_projectAppService.GetAsync(ProjectName)`

File: modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/Index.cshtml [hunk 3]
```diff
@@ -348,14 +349,29 @@
...
+         @(await Component.InvokeAsync<ErrorViewComponent>(new
+          {
+              model = new ErrorPageModel
+              {
+                  RedirectUrl = Model.DocumentsUrlPrefix,
+                  ErrorCode =  "404",
+                  ErrorMessage = L.GetString("ProjectNotFound")
+              }
+          }))
+    }
```
Comment (maliming) [inline modules/docs/src/Volo.Docs.Web/Pages/Documents/Project/Index.cshtml:376]: `EntityFound (project)` should be checked first.

**Comment by ebicoglu** [comment]
image
I tested with an invalid document and still throws 500 exception

**Comment by liangshiw** [comment]
image

HI @ebicoglu
I did the same test, but no problem was found, Maybe it's your configuration problem? like GitHub accessToekn invalid.

**Comment by ebicoglu** [comment]
I can see the other pages. 
I run in Debug mode. Is it a problem?

**Comment by liangshiw** [comment]
Not related to debug mode, Can you provide steps to reproduce?

**Comment by ebicoglu** [comment]
I'm trying with the FileSystem

```
System.IO.FileNotFoundException: Could not find file 'D:\Github\abp\docs\en\Getting-Started23.md'.
File name: 'D:\Github\abp\docs\en\Getting-Started23.md'
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.StreamReader.ValidateArgsAndOpenPath(String path, Encoding encoding, Int32 bufferSize)
   at System.IO.StreamReader..ctor(String path)
   at System.IO.File.OpenText(String path)
   at Volo.Abp.IO.FileHelper.ReadAllTextAsync(String path) in D:\Github\abp\framework\src\Volo.Abp.Core\Volo\Abp\IO\FileHelper.cs:line 57
   at Volo.Docs.FileSystem.Documents.FileSystemDocumentSource.GetDocumentAsync(Project project, String documentName, String languageCode, String version, Nullable`1 lastKnownSignificantUpdateTime) in D:\Github\abp\modules\docs\src\Volo.Docs.Domain\Volo\Docs\FileSystem\Documents\FileSystemDocumentSource.cs:line 28
   at Volo.Docs.Documents.DocumentAppService.GetDocumentAsync(String documentName, Project project, String languageCode, String version, Document oldDocument) in D:\Github\abp\modules\docs\src\Volo.Docs.Application\Volo\Docs\Documents\DocumentAppService.cs:line 257
   at Volo.Docs.Documents.DocumentAppService.GetDocumentWithDetailsDtoAsync(Project project, String documentName, String languageCode, String version) in D:\Github\abp\modules\docs\src\Volo.Docs.Application\Volo\Docs\Documents\DocumentAppService.cs:line 215
   at Volo.Docs.Documents.DocumentAppService.GetAsync(GetDocumentInput input) in D:\Github\abp\modules\docs\src\Volo.Docs.Application\Volo\Docs\Documents\DocumentAppService.cs:line 67
   at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
   at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync() in D:\Github\abp\framework\src\Volo.Abp.Castle.Core\Volo\Abp\Castle\DynamicProxy\CastleAbpMethodInvocationAdapterWithReturnValue.cs:line 24
   at Volo.Abp.Validation.ValidationInterceptor.InterceptAsync(IAbpMethodInvocation invocation) in D:\Github\abp\framework\src\Volo.Abp.Validation\Volo\Abp\Validation\ValidationInterceptor.cs:line 19
   at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed) in D:\Github\abp\framework\src\Volo.Abp.Castle.Core\Volo\Abp\Castle\DynamicProxy\CastleAsyncAbpInterceptorAdapter.cs:line 29
   at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
   at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync() in D:\Github\abp\framework\src\Volo.Abp.Castle.Core\Volo\Abp\Castle\DynamicProxy\CastleAbpMethodInvocationAdapterWithReturnValue.cs:line 24
   at Volo.Abp.Auditing.AuditingInterceptor.InterceptAsync(IAbpMethodInvocation invocation) in D:\Github\abp\framework\src\Volo.Abp.Auditing\Volo\Abp\Auditing\AuditingInterceptor.cs:line 25
   at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed) in D:\Github\abp\framework\src\Volo.Abp.Castle.Core\Volo\Abp\Castle\DynamicProxy\CastleAsyncAbpInterceptorAdapter.cs:line 29
   at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
   at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync() in D:\Github\abp\framework\src\Volo.Abp.Castle.Core\Volo\Abp\Castle\DynamicProxy\CastleAbpMethodInvocationAdapterWithReturnValue.cs:line 24
   at Volo.Abp.Uow.UnitOfWorkInterceptor.InterceptAsync(IAbpMethodInvocation invocation) in D:\Github\abp\framework\src\Volo.Abp.Uow\Volo\Abp\Uow\UnitOfWorkInterceptor.cs:line 31
   at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed) in D:\Github\abp\framework\src\Volo.Abp.Castle.Core\Volo\Abp\Castle\DynamicProxy\CastleAsyncAbpInterceptorAdapter.cs:line 29
   at Volo.Docs.Pages.Documents.Project.IndexModel.SetDocumentAsync() in D:\Github\abp\modules\docs\src\Volo.Docs.Web\Pages\Documents\Project\Index.cshtml.cs:line 363
   at Volo.Docs.Pages.Documents.Project.IndexModel.SetPageAsync() in D:\Github\abp\modules\docs\src\Volo.Docs.Web\Pages\Documents\Project\Index.cshtml.cs:line 152
   at Volo.Docs.Pages.Documents.Project.IndexModel.OnGetAsync() in D:\Github\abp\modules\docs\src\Volo.Docs.Web\Pages\Documents\Project\Index.cshtml.cs:line 101
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.ExecutorFactory.GenericTaskHandlerMethod.Convert[T](Object taskAsObject)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.ExecutorFactory.GenericTaskHandlerMethod.Execute(Object receiver, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeHandlerMethodAsync()
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeNextPageFilterAsync()
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.Rethrow(PageHandlerExecutedContext context)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ExceptionContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.RequestLocalization.AbpRequestLocalizationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) in D:\Github\abp\framework\src\Volo.Abp.AspNetCore\Microsoft\AspNetCore\RequestLocalization\AbpRequestLocalizationMiddleware.cs:line 34
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()
2020-04-27 02:51:14.617 +03:00 [INF] Request finished in 977.0161ms 500 
```

To test it with File System you can use the below configuration in `Projects` table

```
INSERT [dbo].[DocsProjects] ([Id], [ExtraProperties], [ConcurrencyStamp], [Name], [ShortName], [Format], [DefaultDocumentName], [NavigationDocumentName], [ParametersDocumentName], [MinimumVersion], [DocumentStoreType], [MainWebsiteUrl], [LatestVersionBranchName]) VALUES (N'f441f193-0690-e8d3-d56b-39f0c76dd5d5', N'{"Path":"D:\\Github\\abp\\docs"}', N'f745dc26a91c476f9409869b03d08d7f', N'ABP Framework', N'abp', N'md', N'Entities', N'docs-nav.json', N'docs-params.json', NULL, N'FileSystem', N'https://localhost:44326', N'dev')
GO
INSERT [dbo].[DocsProjects] ([Id], [ExtraProperties], [ConcurrencyStamp], [Name], [ShortName], [Format], [DefaultDocumentName], [NavigationDocumentName], [ParametersDocumentName], [MinimumVersion], [DocumentStoreType], [MainWebsiteUrl], [LatestVersionBranchName]) VALUES (N'aec2f637-a766-e6a8-6e96-39f234591167', N'{"Path":"D:\\Github\\abp-commercial-docs"}', N'c75fae127b684a71855bfaf3d74b322d', N'ABP Commercial', N'commercial', N'md', N'Index', N'docs-nav.json', N'docs-params.json', NULL, N'FileSystem', N'https://localhost:44326', N'master')

```

**Comment by liangshiw** [comment]
Hi @ebicoglu,
I updated the code. When the file does not exist, it can be handled well.

**Pull request merged**